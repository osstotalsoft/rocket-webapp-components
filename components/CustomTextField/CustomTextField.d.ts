import { InputAdornmentProps } from "@material-ui/core";
import { StandardTextFieldProps, FilledTextFieldProps, OutlinedTextFieldProps } from "@material-ui/core/TextField";
import React from "react";

export interface Props {
    /**
     * Override or extend the styles applied to the component.
     */
    className?: string
    /**
     * @default false
     * If true, the input will take up the full width of its container.
     */
    fullWidth?: boolean
    /**
     * The variant to use.
     */
    variant?: 'filled' | 'outlined' | 'standard'
    /**
     * If dense or normal, will adjust vertical spacing of this and contained components.
     */
    margin?: 'dense' | 'normal' | 'none'
    /**
     * @default false
     * If true, the label is displayed as required and the input element` will be required.
     */
    required?: boolean
    /**
     * @default false
     * If true, a textarea element will be rendered instead of an input.
     */
    multiline?: boolean
    /**
     * Minimum number of rows to display when multiline option is set to true.
     */
    minRows?: number | string
    /**
     * Maximum number of rows to display when multiline option is set to true.
     */
    maxRows?: number | string
    /**
     * @default false
     * If true, the input will accept only numeric values.
     */
    isNumeric?: boolean
    /**
     * Other properties you can provide to the Input component.
     */
    customInputProps?: Object
    /**
     * Attributes applied to the input element.
     */
    inputProps?: Object
    /**
     * The short hint displayed in the input before the user enters a value.
     */
    placeholder?: string
    /**
     * The adornment of componenent. (Usually an InputAdornment from material-ui)
     */
    endAdornment?: InputAdornmentProps
    /**
     * Start adornment of componenent. (Usually an InputAdornment from material-ui)
     */
    startAdornment: InputAdornmentProps
    /**
     * The label content.
     */
    label?: React.ReactNode
    /**
     * Props applied to the InputLabel element.
     */
    InputLabelProps?: Object
    /**
     * The value of the `input` element, required for a controlled component.
     */
    value?: unknown
    /**
     * The default value of the input element.
     */
    defaultValue?: unknown
    /**
     * @default false
     * If true, the input element will be disabled.
     */
    disabled?: boolean
    /**
     * @default false
     * If true, the label will be displayed in an error state.
     */
    error?: boolean
    /**
     * The helper text content.
     */
    helperText?: React.ReactNode
    /**
     * Callback fired when the value is changed.
     * @default  () => { }
     * @param {object} event The event source of the callback.
     * You can pull out the new value by accessing `event.target.value` (string).
     */
    onChange?: (object: Object) => void
    /**
     * @default 0
     * The delay of debouncing.
     */
    debounceBy?: number
    /**
     * The current language, preferably taken from the i18next (i18.language) or another internationalization library
     */
    language: string
}

export type CustomTextFieldProps = Props | StandardTextFieldProps | FilledTextFieldProps | OutlinedTextFieldProps;

/**
 *
 * Demos:
 *
 * - https://bit.dev/totalsoft_oss/react-mui/custom-text-field
 *
 *  Take a look over the code for further information:https://github.com/osstotalsoft/rocket-webapp-components/blob/master/components/CustomTextField/CustomTextField.js
 * 
 * Material-UI component used:
 * 
 * - TextField: https://material-ui.com/api/text-field/
 * 
 */
export default function CustomTextField(props: CustomTextFieldProps): JSX.Element;