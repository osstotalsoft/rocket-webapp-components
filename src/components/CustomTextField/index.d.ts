export { default } from './CustomTextField';
export * from './CustomTextField';