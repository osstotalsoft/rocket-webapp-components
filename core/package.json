{
  "engines": {
    "node": ">=10"
  },
  "scripts": {
    "prebuild": "rimraf build",
    "build": "npm run build:cjs &&  npm run build:copy-files",
    "build:cjs": "babel src -d build --copy-files",
    "build:copy-files": "node ./scripts/copy-files.js",
    "bit-build": "bit -v && bit import && bit build"
  },
  "devDependencies": {
    "@babel/cli": "^7.8.0",
    "@babel/core": "^7.7.7",
    "@babel/node": "^7.7.7",
    "@babel/plugin-proposal-decorators": "^7.8.3",
    "@babel/preset-env": "^7.7.7",
    "@babel/preset-react": "^7.8.0",
    "babel-eslint": "^10.0.3",
    "cross-env": "^7.0.2",
    "eslint": "^6.8.0",
    "eslint-config-prettier": "^6.7.0",
    "eslint-config-standard": "^14.1.0",
    "eslint-config-standard-react": "^9.2.0",
    "eslint-plugin-import": "^2.18.2",
    "eslint-plugin-node": "^11.0.0",
    "eslint-plugin-prettier": "^3.1.1",
    "eslint-plugin-promise": "^4.2.1",
    "eslint-plugin-react": "^7.17.0",
    "eslint-plugin-standard": "^4.0.1",
    "gh-pages": "^2.2.0",
    "npm-run-all": "^4.1.5",
    "prettier": "^1.19.1",
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "react-scripts": "^3.4.1",
    "rimraf": "^3.0.0",
    "url-search-params-polyfill": "^7.0.1"
  },
  "peerDependencies": {
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "react-flags": "^0.1.18",
    "react-router-dom": "^5.1.2"
  },
  "sideEffects": false,
  "dependencies": {},
  "bit": {
    "env": {
      "compiler": "totalsoft_oss.react-mui/compilers/react@1.0.32"
    },
    "componentsDefaultDirectory": "components/{name}",
    "packageManager": "npm"
  }
}
