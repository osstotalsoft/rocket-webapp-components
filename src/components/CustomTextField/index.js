import CustomTextField from "./CustomTextField";
export default CustomTextField;