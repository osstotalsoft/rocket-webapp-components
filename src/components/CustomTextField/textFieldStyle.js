import styles from '../../assets/jss/styles';

const textFieldStyle = theme => {
    const { defaultFont, grayColor } = styles(theme);
    return {
        label: {
            ...defaultFont,
            color: grayColor[3] + " !important",
            fontWeight: "400",
            lineHeight: "1.42857",
            top: "0px",
            marginTop: "0px"
        },
        input: {
            color: grayColor[14],
            "&,&::placeholder": {
                ...defaultFont,
                fontWeight: "400",
                opacity: "1"
            },
            "&::placeholder": {
                color: grayColor[3]
            }
        },
    };
}

export default textFieldStyle;