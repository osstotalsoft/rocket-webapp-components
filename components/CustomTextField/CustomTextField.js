/* eslint-disable react/prop-types */
import React, { useState, useEffect, useCallback } from "react";
import PropTypes from "prop-types";
import NumberFormat from "react-number-format";
import { makeStyles, TextField } from "@material-ui/core";
import textFieldStyle from "./textFieldStyle";
import { useDebounce } from "use-debounce";
import { curry } from "ramda";
import { emptyFunction, emptyString } from "../../utils/constants";

const useStyles = makeStyles(textFieldStyle);

const isAllowed = values =>
  !(values.floatValue && values.floatValue.toString().includes("e"));

function NumberFormatCustom(props) {
  const {
    inputRef,
    onChange,
    decimalScale,
    fixedDecimalScale,
    thousandSeparator,
    language,
    ...other
  } = props;
  let formatter = new Intl.NumberFormat(language);

  var thousandSep = thousandSeparator
    ? formatter.format(1111).replace(/1/g, "")
    : thousandSeparator;
  var decimalSeparator = formatter.format(1.1).replace(/1/g, "");
  const valueIsNumericString =
    !!props.value &&
    typeof props.value === "string" &&
    !Number.isNaN(Number(props.value));

  const handleValueChange = useCallback(values => onChange(values.floatValue), [
    onChange
  ]);

  return (
    <NumberFormat
      {...other}
      getInputRef={inputRef}
      style={{ textAlign: "right" }}
      onValueChange={handleValueChange}
      isAllowed={isAllowed}
      decimalScale={decimalScale}
      fixedDecimalScale={fixedDecimalScale}
      thousandSeparator={thousandSep}
      decimalSeparator={decimalSeparator}
      isNumericString={valueIsNumericString}
      //format = {customFormat}
    />
  );
}

NumberFormatCustom.propTypes = {
  inputRef: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  decimalScale: PropTypes.number.isRequired,
  fixedDecimalScale: PropTypes.bool,
  thousandSeparator: PropTypes.oneOfType([PropTypes.string, PropTypes.bool])
};

NumberFormatCustom.defaultProps = {
  decimalScale: 2,
  fixedDecimalScale: true,
  thousandSeparator: true
};

// TODO: this is for backwards compatibility (onTextBoxChange) ... remove onTextBoxChange and keep only value
const setValueToDesiredFormat = curry((isNumeric, value) =>
  isNumeric ? value : { target: { value } }
);
const getValueFromDesiredFormat = curry((isNumeric, e) =>
  isNumeric ? e : e?.target.value
);

function CustomTextField({
  isNumeric,
  customInputProps,
  inputProps,
  endAdornment,
  startAdornment,
  InputLabelProps,
  className,
  value,
  onChange,
  debounceBy,
  ...rest
}) {
  const classes = useStyles();

  // props applied to the Input element
  const customMuiInput = isNumeric
    ? {
        inputComponent: NumberFormatCustom,
        endAdornment,
        startAdornment,
        ...customInputProps
      }
    : {
        endAdornment,
        startAdornment,
        ...customInputProps
      };

  // attributes applied to the input element
  const customReactInput = {
    ...inputProps,
    className: `${classes.input} ${
      inputProps?.className ? inputProps.className : emptyString
    }`
  };

  const [localValue, setLocalValue] = useState(value);
  const [debouncedValue] = useDebounce(localValue, debounceBy);

  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(() => {
    debouncedValue !== value &&
      onChange(setValueToDesiredFormat(isNumeric)(debouncedValue));
  }, [debouncedValue]);
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(() => {
    value !== localValue && setLocalValue(value);
  }, [value]);

  const handleChange = useCallback(
    e => setLocalValue(getValueFromDesiredFormat(isNumeric)(e)),
    [isNumeric]
  );

  return (
    <TextField
      onChange={handleChange}
      value={localValue}
      {...rest}
      className={classes.textField + " " + className}
      InputProps={customMuiInput}
      inputProps={customReactInput}
      InputLabelProps={{
        className: classes.label,
        ...InputLabelProps
      }}
    />
  );
}

CustomTextField.defaultProps = {
  isNumeric: false,
  debounceBy: 0,
  onChange: emptyFunction
};

CustomTextField.propTypes = {
  /**
   * Override or extend the styles applied to the component.
   */
  className: PropTypes.string,
  /**
   * If true, the input will accept only numeric values.
   */
  isNumeric: PropTypes.bool,
  /**
   * Other properties you can provide to the Input component.
   */
  customInputProps: PropTypes.object,
  /**
   * Attributes applied to the input element.
   */
  inputProps: PropTypes.object,
  /**
   * End adornment of componenent. (Usually an InputAdornment from material-ui)
   */
  endAdornment: PropTypes.any,
  /**
   * Start adornment of componenent. (Usually an InputAdornment from material-ui)
   */
  startAdornment: PropTypes.any,
  /**
   * Props applied to the InputLabel element.
   */
  InputLabelProps: PropTypes.object,
  /**
   * The value of the `input` element, required for a controlled component.
   */
  value: PropTypes.any,
  /**
   * Callback fired when the value is changed.
   *
   * @param {object} event The event source of the callback.
   * You can pull out the new value by accessing `event.target.value` (string).
   */
  onChange: PropTypes.func,
  /**
   * The delay of debouncing.
   */
  debounceBy: PropTypes.number,
  /**
   * The current language, preferably taken from the i18next (i18.language) or another internationalization library
   */
  language: PropTypes.string
};

export default CustomTextField;
