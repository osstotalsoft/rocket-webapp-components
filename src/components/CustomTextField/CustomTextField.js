/* eslint-disable react/prop-types */
import React, { useState, useEffect, useCallback } from "react";
import PropTypes from "prop-types";
import NumberFormat from "react-number-format";
import { makeStyles, TextField, deprecatedPropType } from "@material-ui/core";
import textFieldStyle from "./textFieldStyle";
import { useDebounce } from "use-debounce";
import { curry, is, head } from "ramda";
import { emptyFunction, emptyString } from "../../utils/constants";

const useStyles = makeStyles(textFieldStyle);

const isAllowed = values =>
  !(values.floatValue && values.floatValue.toString().includes("e"));

function NumberFormatCustom(props) {
  const {
    inputRef,
    onChange,
    decimalScale,
    fixedDecimalScale,
    thousandSeparator,
    language,
    ...other
  } = props;
  let formatter = new Intl.NumberFormat(language);

  var thousandSep = thousandSeparator
    ? formatter.format(1111).replace(/1/g, "")
    : thousandSeparator;
  var decimalSeparator = formatter.format(1.1).replace(/1/g, "");
  const valueIsNumericString =
    !!props.value &&
    typeof props.value === "string" &&
    !Number.isNaN(Number(props.value));

  const handleValueChange = useCallback(values => onChange(values.floatValue), [
    onChange
  ]);

  return (
    <NumberFormat
      {...other}
      getInputRef={inputRef}
      style={{ textAlign: "right" }}
      onValueChange={handleValueChange}
      isAllowed={isAllowed}
      decimalScale={decimalScale}
      fixedDecimalScale={fixedDecimalScale}
      thousandSeparator={thousandSep}
      decimalSeparator={decimalSeparator}
      isNumericString={valueIsNumericString}
      //format = {customFormat}
    />
  );
}

NumberFormatCustom.propTypes = {
  inputRef: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  decimalScale: PropTypes.number.isRequired,
  fixedDecimalScale: PropTypes.bool,
  thousandSeparator: PropTypes.oneOfType([PropTypes.string, PropTypes.bool])
};

NumberFormatCustom.defaultProps = {
  decimalScale: 2,
  fixedDecimalScale: true,
  thousandSeparator: true
};

// TODO: this is for backwards compatibility (onTextBoxChange) ... remove onTextBoxChange and keep only value
const setValueToDesiredFormat = curry((isNumeric, value) =>
  isNumeric ? value : { target: { value } }
);
const getValueFromDesiredFormat = curry((isNumeric, e) =>
  isNumeric ? e : e?.target.value
);

function CustomTextField({
  isNumeric,
  customInputProps,
  inputProps,
  InputProps,
  endAdornment,
  startAdornment,
  InputLabelProps,
  className,
  value,
  onChange,
  debounceBy,
  helperText,
  ...rest
}) {
  const classes = useStyles();

  // props applied to the Input element
  const customMuiInput = isNumeric
    ? {
        inputComponent: NumberFormatCustom,
        endAdornment,
        startAdornment
      }
    : {
        endAdornment,
        startAdornment,
        ...InputProps
      };

  // attributes applied to the input element
  const customReactInput = {
    ...inputProps,
    className: `${classes.input} ${
      inputProps?.className ? inputProps.className : emptyString
    }`
  };

  const [localValue, setLocalValue] = useState(value);
  const [debouncedValue] = useDebounce(localValue, debounceBy);

  useEffect(() => {
    debouncedValue !== value &&
      onChange(setValueToDesiredFormat(isNumeric)(debouncedValue));
  }, [debouncedValue]); // eslint-disable-line react-hooks/exhaustive-deps

  useEffect(() => {
    value !== localValue && setLocalValue(value);
  }, [value]); // eslint-disable-line react-hooks/exhaustive-deps

  const handleChange = useCallback(
    e => setLocalValue(getValueFromDesiredFormat(isNumeric)(e)),
    [isNumeric]
  );

  return (
    <TextField
      onChange={handleChange}
      value={localValue}
      helperText={is(Array, helperText) ? head(helperText) : helperText}
      {...rest}
      className={className ? `${classes.textField} ${className}` : classes.textField}
      InputProps={customMuiInput}
      inputProps={customReactInput}
      InputLabelProps={{
        className: classes.label,
        ...InputLabelProps
      }}
    />
  );
}

CustomTextField.defaultProps = {
  isNumeric: false,
  debounceBy: 0,
  onChange: emptyFunction
};

CustomTextField.propTypes = {
  /**
   * Override or extend the styles applied to the component.
   */
  className: PropTypes.string,
  /**
   * If true, the input will accept only numeric values.
   */
  isNumeric: PropTypes.bool,
  /**
   * Other properties you can provide to the component.
   */
  customInputProps: deprecatedPropType(PropTypes.object, "Use `inputProps` instead."),
  /**
   * Attributes applied to the input element.
   * For the numeric input, you can provide properties like thousandSeparator, decimalScale and allowNegative.
   */
  inputProps: PropTypes.object,
  /**
   * Other properties you can provide to the Input component.
   */
  InputProps: PropTypes.object,
  /**
   * End adornment of componenent. (Usually an InputAdornment from material-ui)
   */
  endAdornment: PropTypes.any,
  /**
   * Start adornment of componenent. (Usually an InputAdornment from material-ui)
   */
  startAdornment: PropTypes.any,
  /**
   * Props applied to the InputLabel element.
   */
  InputLabelProps: PropTypes.object,
  /**
   * The value of the `input` element, required for a controlled component.
   */
  value: PropTypes.any,
  /**
   * The content of the helper under the input.
   */
  helperText: PropTypes.node,
  /**
   * Callback fired when the value is changed.
   *
   * @param {object} event The event source of the callback.
   * You can pull out the new value by accessing `event.target.value` (string).
   */
  onChange: PropTypes.func,
  /**
   * The delay of debouncing.
   */
  debounceBy: PropTypes.number,
  /**
   * The current language, preferably taken from the i18next (i18.language) or another internationalization library
   */
  language: PropTypes.string
};

export default CustomTextField;
