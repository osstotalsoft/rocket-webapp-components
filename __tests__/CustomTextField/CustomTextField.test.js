import React from 'react';
import renderer from 'react-test-renderer';
import CustomTextField from '../../components/CustomTextField/CustomTextField';
import { mount } from 'enzyme';
import { TextField } from "@material-ui/core";

describe('CustomTextField', () => {
    it('snapshot', () => {
        const onCustomInputChange = jest.fn()

        const component = renderer.create(<CustomTextField
            fullWidth
            label={'CustomTextField'}
            value={12 || ""}
            onChange={onCustomInputChange}
        />);

        const json = component.toJSON();
        expect(json).toMatchSnapshot();
    });

    it('if isNumeric is true then inputComponent node of type NumberFormatCustom exists', () => {
        const onCustomInputChange = jest.fn()

        const wrapper = mount(<CustomTextField
            fullWidth
            label={'CustomTextField'}
            value={12 || ""}
            onChange={onCustomInputChange}
            inputProps={{
                decimalScale: 3,
                thousandSeparator: true,
                prefix: '$'
            }}
            isNumeric
        />);

        console.log(wrapper.find(TextField).props())

        expect(wrapper.find(TextField).props().InputProps.inputComponent).not.toBeUndefined()
    });

    it('if isNumeric is false then inputComponent node of type NumberFormatCustom does not exist', () => {
        const onCustomInputChange = jest.fn()

        const wrapper = mount(<CustomTextField
            fullWidth
            label={'CustomTextField'}
            value={12 || ""}
            onChange={onCustomInputChange}
        />);

        expect(wrapper.find(TextField).props().InputProps.inputComponent).toBeUndefined()
    });

    it('inputProps class input is present', () => {
        const onCustomInputChange = jest.fn()

        const wrapper = mount(<CustomTextField
            fullWidth
            label={'CustomTextField'}
            value={12 || ""}
            onChange={onCustomInputChange}
        />);

        expect(wrapper.find(TextField).props().inputProps.className).toContain('input')
    });
});