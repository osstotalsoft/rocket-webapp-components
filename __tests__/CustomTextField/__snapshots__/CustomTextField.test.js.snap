// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomTextField snapshot 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
>
  <label
    className="MuiFormLabel-root MuiInputLabel-root makeStyles-label-1 MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
    data-shrink={true}
  >
    CustomTextField
  </label>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiInputBase-input MuiInput-input makeStyles-input-2 "
      disabled={false}
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="text"
      value={12}
    />
  </div>
</div>
`;
